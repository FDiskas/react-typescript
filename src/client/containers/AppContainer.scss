.appContainer {
    padding: 5px;
    margin: 10px;
    outline: 5px solid #fff;
    background-color: rgba(255, 194, 177, 0.51);
}
