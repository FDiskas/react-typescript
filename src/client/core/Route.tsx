import * as React from 'react';
import { Route as ReactRoute, RouteProps } from 'react-router-dom';

export interface IRouteProps {
    layout?: React.ComponentType<RouteProps> | React.ComponentType<any>;
    component: React.ComponentType<RouteProps> | React.ComponentType<any>;
}

export interface IRouteState {
}

export class Route extends React.Component<IRouteProps & RouteProps, IRouteState> {
    constructor(props: IRouteProps) {
        super(props);
    }

    render() {
        const { component, layout, ...rest } = this.props;
        let routeComponent = (props: any) => React.createElement(component, props);

        if (layout) {
            routeComponent = props =>
                React.createElement(layout, props, React.createElement(component, props));
        }

        return <ReactRoute {...rest} render={routeComponent} />;
    }
}
