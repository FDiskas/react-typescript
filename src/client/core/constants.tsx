export const ACTION = {
    CHECKBOX: {
        ADD: 'ac@checkbox.add',
        DELETE: 'ac@checkbox.delete',
    },
};
