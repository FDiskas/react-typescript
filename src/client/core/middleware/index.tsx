import reduxLogger from 'redux-logger';
export { reduxLogger };
