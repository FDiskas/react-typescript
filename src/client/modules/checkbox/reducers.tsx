import { Action, handleActions } from 'redux-actions';
import { IItem } from '../../components/checkbox/CheckBoxComponent';
import { ACTION } from '../../core/constants';

const initialState: IItem[] = [];

export default handleActions(
    {
        [ACTION.CHECKBOX.ADD]: (state: IItem[], action: Action<IItem>) => {
            return [...state, action.payload];
        },
        [ACTION.CHECKBOX.DELETE]: (state: IItem[], action: Action<IItem>) => {
            return state.filter(v => v !== action.payload);
        },
    },

    initialState,
);
