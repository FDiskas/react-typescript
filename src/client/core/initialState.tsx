import { IItem } from '../components/checkbox/CheckBoxComponent';

export const initialState = {
    checkBoxItems: [
        { id: '1', label: 'Label 1', value: 'Value 1' },
        { id: '2', label: 'Label 2', value: 'Value 2' },
    ] as IItem[],
};
