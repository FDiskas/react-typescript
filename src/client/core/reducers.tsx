import { combineReducers } from 'redux';

import reducers from '../modules/checkbox/reducers';
import { IItem } from '../components/checkbox/CheckBoxComponent';

export interface IRootState {
  checkBoxItems: IItem[];
}

const rootReducer = combineReducers({
  checkBoxItems: reducers,
});

export default rootReducer;
