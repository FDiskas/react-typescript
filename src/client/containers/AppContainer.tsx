import * as React from 'react';
import { bindActionCreators } from 'redux';
import { connect, MapDispatchToProps } from 'react-redux';
import * as uniqId from 'uniqid';

import { CheckBox, IItem } from '../components/checkbox/CheckBoxComponent';
import { addCheckbox } from '../modules/checkbox/actions';
import { IRootState } from '../core/reducers';

export interface IAppProps {
    addAction: typeof addCheckbox;
    items: IItem[];
}

export interface IAppState {
    items: IItem[];
}

class App extends React.Component<IAppProps, IAppState> {
    static defaultProps: Partial<IAppProps> = {
        items: [{
            id: 'string',
            label: 'Demo',
            value: 'sadasd',
        }],

        addAction: addCheckbox,
    };

    selectedItems = new Map();

    constructor(props: IAppProps) {
        super(props);

        this.state = { items: [] };
    }

    handleClick = (item: IItem, checked: boolean) => {
        if (checked) {
            this.selectedItems.set(item.id, item);
        } else {
            this.selectedItems.delete(item.id);
        }
    };

    isSelected(item: IItem): boolean {
        return this.selectedItems.has(item.id);
    }

    handleAdd = () => {
        this.props.addAction({
            id: uniqId(),
            value: 'demo',
            label: 'Demo',
        });
    };

    handleDelete = () => {
        // console.log(item);
    };

    render() {
        return (
            <div>
                <button onClick={this.handleAdd}>Add more</button>
                {this.props.items.map(item => (
                    <div key={item.id}>
                        <CheckBox
                            item={item}
                            handleClick={this.handleClick}
                        />
                        <button onClick={this.handleDelete.bind(this, item)}>Delete</button>
                    </div>
                ))}
            </div>
        );
    }
}

export function mapStateToProps(state: IRootState) {
    return ({
        items: state.checkBoxItems,
    });
}

export function mapDispatchToProps(dispatch: MapDispatchToProps<any, any>) {
    return ({
        addAction: bindActionCreators(addCheckbox, dispatch),
    });
}

export default connect(mapStateToProps, mapDispatchToProps)(App);

export { App as AppDisconnected };
