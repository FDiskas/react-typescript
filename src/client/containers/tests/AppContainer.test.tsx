import * as React from 'react';
import { shallow } from 'enzyme';

import { AppDisconnected, mapStateToProps } from '../AppContainer';
import { IRootState } from '../../core/reducers';

describe('AppContainer.tsx', () => {
    const props = {
        items: [
            { id: 'demo1', label: 'Demo', value: 'demo' },
            { id: 'demo2', label: 'Demo2', value: 'demo2' },
        ],
        addAction: jest.fn(),
    };

    it('renders without crashing', () => {
        shallow(<AppDisconnected {...props} />);
    });

    it('matches snapshot', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('Buttons should be visible', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        expect(wrapper.find('button').length).toBe(3);
    });

    it('should handle click', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        const instance = wrapper.instance() as AppDisconnected;
        const params = { id: 'test', value: 1, label: 'Demo' };

        instance.handleClick(params, true);
        expect(instance.selectedItems.has('test')).toBeTruthy();

        instance.handleClick(params, false);
        expect(instance.selectedItems.has('test')).toBeFalsy();
    });

    it('handleAdd()', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        const instance = wrapper.instance() as AppDisconnected;

        instance.handleAdd();

        expect(props.addAction).toBeCalledWith({ id: expect.any(String), label: 'Demo', value: 'demo' });
    });

    it('mapStateToProps()', () => {
        const initialState: IRootState = {
            checkBoxItems: [],
        };

        expect(mapStateToProps(initialState)).toEqual({ items: [] });
    });

    it('Add to selected list and check if exists', () => {
        const wrapper = shallow(<AppDisconnected {...props} />);
        const instance = wrapper.instance() as AppDisconnected;
        const item = { id: 'test', value: 1, label: 'Demo' };

        instance.selectedItems.set(item.id, item);

        expect(instance.isSelected(item)).toBe(true);
    })
});
