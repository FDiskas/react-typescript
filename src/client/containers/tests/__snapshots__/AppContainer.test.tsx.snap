// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer.tsx matches snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Add more
  </button>
  <div
    key="demo1"
  >
    <CheckBox
      checked={false}
      handleClick={[Function]}
      item={
        Object {
          "id": "demo1",
          "label": "Demo",
          "value": "demo",
        }
      }
    />
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
